pkg/controller/direct/certificatemanager/dnsauthorization_controller.go

```diff
@@ -0,0 +1,334 @@
+	log := klog.FromContext(ctx).WithName(ctrlName)
+	log.V(2).Info("getting CertificateManagerDNSAuthorization", "name", a.id.FullyQualifiedName())
+
+	req := &certificatemanagerpb.GetDnsAuthorizationRequest{Name: a.id.FullyQualifiedName()}
+	dnsauthorizationpb, err := a.gcpClient.GetDnsAuthorization(ctx, req)
+	if err != nil {
+		if direct.IsNotFound(err) {
+			return false, nil
+		}
+		return false, fmt.Errorf("getting CertificateManagerDNSAuthorization %q: %w", a.id.FullyQualifiedName(), err)
+	}
+
+	a.actual = dnsauthorizationpb
+	return true, nil
+}
+
+func (a *Adapter) Create(ctx context.Context, createOp *directbase.CreateOperation) error {
+	u := createOp.GetUnstructured()
+
+	log := klog.FromContext(ctx).WithName(ctrlName)
+	log.V(2).Info("creating DnsAuthorization", "name", a.id.FullyQualifiedName())
+	mapCtx := &direct.MapContext{}
+
+	desired := a.desired.DeepCopy()
+	resource := CertificateManagerDNSAuthorizationSpec_ToProto(mapCtx, &desired.Spec)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+	resource.Labels = make(map[string]string)
+	for k, v := range a.desired.GetObjectMeta().GetLabels() {
+		resource.Labels[k] = v
+	}
+
+	req := &certificatemanagerpb.CreateDnsAuthorizationRequest{
+		Parent:             a.id.Parent.String(),
+		DnsAuthorizationId: a.id.DnsAuthorization,
+		DnsAuthorization:   resource,
+	}
+	op, err := a.gcpClient.CreateDnsAuthorization(ctx, req)
+	if err != nil {
+		return fmt.Errorf("creating DnsAuthorization %s: %w", a.id.FullyQualifiedName(), err)
+	}
+	created, err := op.Wait(ctx)
+	if err != nil {
+		return fmt.Errorf("DnsAuthorization %s waiting creation: %w", a.id.FullyQualifiedName(), err)
+	}
+	log.V(2).Info("successfully created DnsAuthorization", "name", a.id.FullyQualifiedName())
+
+	status := &krm.CertificateManagerDNSAuthorizationStatus{}
+	status = CertificateManagerDNSAuthorizationStatusObservedState_FromProto(mapCtx, created)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+	return setStatus(u, status)
+}
+
+func (a *Adapter) Update(ctx context.Context, updateOp *directbase.UpdateOperation) error {
+	u := updateOp.GetUnstructured()
+
+	log := klog.FromContext(ctx).WithName(ctrlName)
+	log.V(2).Info("updating DnsAuthorization", "name", a.id.FullyQualifiedName())
+	mapCtx := &direct.MapContext{}
+	updateMask := &fieldmaskpb.FieldMask{}
+
+	if !reflect.DeepEqual(a.desired.Spec.Description, a.actual.Description) {
+		updateMask.Paths = append(updateMask.Paths, "description")
+	}
+
+	if !reflect.DeepEqual(a.desired.Labels, a.actual.Labels) {
+		updateMask.Paths = append(updateMask.Paths, "labels")
+	}
+
+	desired := a.desired.DeepCopy()
+	resource := CertificateManagerDNSAuthorizationSpec_ToProto(mapCtx, &desired.Spec)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+
+	resource.Labels = make(map[string]string)
+	for k, v := range a.desired.GetObjectMeta().GetLabels() {
+		resource.Labels[k] = v
+	}
+
+	req := &certificatemanagerpb.UpdateDnsAuthorizationRequest{
```

Comment:
I just updated the template, I think we want to skip if the updateMask is empty (to save the GCP update calls)

Reply:
done